designs/s3/20200210-design-vscode-s3.md

```diff
@@ -0,0 +1,146 @@
+        default location for future downloads.
+      - The default download location is also configurable as a global Toolkit
+        option. This is not S3-specific, it is used for all "Download"
+        experiences in the Toolkit.
+
+### Drag-drop from client to server
+
+- Drag a folder from the local system to the S3 tree:
+  - onto a bucket uploads as a new top-level folder in the bucket
+  - onto a folder uploads as a new subfolder
+- Drag a file from the local system to the S3 tree:
+  - onto a bucket uploads the file as a top-level object in the bucket
+  - onto a folder uploads the file as an object in the folder
+
+### Drag-drop server-side
+
+- User can copy objects or folders from one folder to another by drag-drop.
+- User can copy objects or folders _across regions_ by dragging from one region
+  to another.
+
+### S3 bucket properties
+
+Surface popular bucket properties such as "public accessible" and "versioning
+enabled" in a clear yet uncluttered manner.
+
+- Show icon (same as AWS web console) next to bucket name indicating whether it
+  is publicly accessible
+- Show icon or sigil next to bucket name indicating whether versioning is enabled
+
+### S3 object properties
+
+S3 objects have properties like size, last-modified-date, and previous versions
+(if versioning is enabled).
+
+- If >=1 previous versions are available for an object, show the count of
+  previous versions next to the object.
+- If >=1 previous versions are available for an object, context menu
+  (right-click) on the object exposes a "Show previous versions" menu item
+  (otherwise greyed-out).
+  - Previous versions are presented as children of the object.
+- Double-clicking an object always performs the "Edit" action, even if the
+  object has previous versions.
+
+## Out of scope
+
+Some features are out of scope for the phase 1 implementation.
+
+* "Moving" objects or folders. To "move" an object, user must copy the object,
+  then delete the original location.
+* List/browser view.  VSCode is a text editor, so creating "inner platforms"
+  should be avoided.  The tree/explorer in the left-column is the idiomatic way
+  to explore data in VSCode. We should see how much leverage we can get out of
+  that common interface before introducing bespoke UIs.
+* Pinning ([ref](https://github.com/aws/aws-toolkit-jetbrains/issues/90)):
+  ability to save AWS constructs to a "recent" or "favorites" area is useful,
+  but requires a holistic treatment: it makes sense for any AWS object, not
+  only S3.
+* Drag-drop from S3 tree view to local system. This could present unpleasant
+  edge-cases depending on the drop-target (e.g. Microsoft Outlook), so it is
+  out of scope for this initial proposal.
+  * The low-friction "Download" workflow described in the proposal addresses
+    the common case, thus drag-drop is nice-to-have.
+
+## Unknowns
+
+* VSCode may offer a rich UI component framework in the future.
```

Comment:
Any elaboration on how this changes the proposed S3 explorer? If you have a vision for something that we can act on quickly, we shouldn't necessarily wait for a feature which has no guarantee of existing.

Reply:
I'm in favor of avoiding this direction entirely, and focusing on the strengths that VSCode gives us today.

Reply:
I agree, I like the simplicity of this approach. I'd say don't mention this if you have no intention to use it in this or future iterations (especially since there's zero indication they'll provide a framework). Otherwise, please provide a way that this might be used, since we may find implementing said features independently valuable.